### PR TITLE
Foundation

### PR DESCRIPTION
**Pull request by loganwright**
14x faster serialization in my perf. tests ... (including on linux)

Foundation:
https://travis-ci.org/vapor/json/jobs/211792443#L259
Jay:
https://travis-ci.org/vapor/json/jobs/211792443#L255

### CODE REVIEW

File: Sources/JSON/Array+JSONConvertible.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+extension Array: JSONConvertible {
+    public init(json: JSON) throws {
+        guard Element.self is JSONInitializable.Type else {
```
Comment (tanner0101) [inline Sources/JSON/Array+JSONConvertible.swift]: you can use `guard let` here instead of force cast

File: Sources/JSON/JSON+Parse.swift [hunk 2]
```diff
@@ -1,35 +1,131 @@
...
-public enum JSONError: Swift.Error {
-    case allowFragmentsNotSupported
-    case parse(path: String, error: Swift.Error)
-}
 
 extension JSON {
     public init(
-        serialized: BytesRepresentable,
-        allowComments: Bool = false,
+        bytes: BytesRepresentable,
```
Comment (tanner0101) [inline Sources/JSON/JSON+Parse.swift]: could be `_ bytes` since `BytesRepresentable` is a strong type

File: Sources/JSON/JSON+Parse.swift [hunk 3]
```diff
@@ -1,35 +1,131 @@
...
+        self = JSON(structuredData)
+    }
+}
+
+extension StructuredData {
+    /// Attempt to initialize a node with a foundation object.
+    ///
+    /// - parameter any: the object to create a node from
+    /// - throws: if fails to create node.
+    public init(json: Any) throws {
```
Comment (tanner0101) [inline Sources/JSON/JSON+Parse.swift]: should this be public? the `Any` stuff is kind of hacky

**Review by tanner0101 — APPROVED** [review]
woooooot!

File: Sources/JSON/Array+JSONConvertible.swift (hunk 1 above)
Comment (loganwright) [reply Sources/JSON/Array+JSONConvertible.swift]: @tanner0101 guard let was crashing for me occasionally on linux, this is a work around adapted from Node.

Can't recall if it was present on 3.1 though, testing now

File: Sources/JSON/JSON+Parse.swift (hunk 3 above)
Comment (loganwright) [reply Sources/JSON/JSON+Parse.swift]: I'll keep it internal, we can open it up later if there's use

File: Sources/JSON/JSON+Parse.swift (hunk 2 above)
Comment (loganwright) [reply Sources/JSON/JSON+Parse.swift]: I think parity with the actual bytes function makes discoverability easier here

**Pull request merged**